File: orbit-bench/src/orbitbench/metrics.py
"""Latency statistics: percentiles over raw samples, means only as context."""

import math
from dataclasses import dataclass, field


def percentile(samples: list[float], q: float) -> float:
    """Linear-interpolated percentile on sorted samples. 0.0 when empty."""
    if not samples:
        return 0.0
    ordered = sorted(samples)
    if len(ordered) == 1:
        return ordered[0]
    k = (len(ordered) - 1) * q
    f = math.floor(k)
    c = math.ceil(k)
    if f == c:
        return ordered[int(k)]
    return ordered[f] + (ordered[c] - ordered[f]) * (k - f)


@dataclass
class Stats:
    backend: str
    scenario: str
    ops: int = 0
    errors: int = 0
    reads: int = 0
    hits: int = 0
    total_s: float = 0.0
    latencies_ms: list[float] = field(default_factory=list, repr=False)

    @property
    def mean_ms(self) -> float:
        return round(sum(self.latencies_ms) / len(self.latencies_ms), 3) if self.latencies_ms else 0.0

    @property
    def p50_ms(self) -> float:
        return round(percentile(self.latencies_ms, 0.50), 3)

    @property
    def p95_ms(self) -> float:
        return round(percentile(self.latencies_ms, 0.95), 3)

    @property
    def p99_ms(self) -> float:
        return round(percentile(self.latencies_ms, 0.99), 3)

    @property
    def ops_per_sec(self) -> float:
        return round(self.ops / self.total_s, 1) if self.total_s > 0 else 0.0

    @property
    def hit_ratio(self) -> float:
        return round(self.hits / self.reads, 3) if self.reads > 0 else 0.0

    def as_dict(self) -> dict:
        return {
            "backend": self.backend,
            "scenario": self.scenario,
            "ops": self.ops,
            "errors": self.errors,
            "reads": self.reads,
            "hit_ratio": self.hit_ratio,
            "total_s": round(self.total_s, 3),
            "ops_per_sec": self.ops_per_sec,
            "mean_ms": self.mean_ms,
            "p50_ms": self.p50_ms,
            "p95_ms": self.p95_ms,
            "p99_ms": self.p99_ms,
        }


def console_table(stats_list: list[Stats]) -> str:
    """Aligned ASCII table - the deliverable of every run."""
    headers = ["backend", "scenario", "ops", "errors", "hit%", "ops/s",
               "mean", "p50", "p95", "p99"]
    rows = []
    for s in stats_list:
        rows.append([
            s.backend, s.scenario, str(s.ops), str(s.errors), f"{s.hit_ratio:.0%}",
            f"{s.ops_per_sec:.0f}", f"{s.mean_ms:.3f}", f"{s.p50_ms:.3f}",
            f"{s.p95_ms:.3f}", f"{s.p99_ms:.3f}",
        ])

    widths = [max(len(h), *(len(r[i]) for r in rows)) if rows else len(h)
              for i, h in enumerate(headers)]
    line = "  ".join(h.ljust(w) for h, w in zip(headers, widths))
    separator = "-" * len(line)
    body = "\n".join("  ".join(cell.ljust(w) for cell, w in zip(row, widths)) for row in rows)
    units = "(latency columns in ms; in-memory backends legitimately show sub-ms values)"
    return f"{line}\n{separator}\n{body}\n{separator}\n{units}"

File: twin-rs/src/metrics.rs
//! Stats - same percentile interpolation and same JSON keys as OrbitBench,
//! so tables from both languages can sit in one document.

pub fn percentile(samples: &mut [f64], q: f64) -> f64 {
    if samples.is_empty() {
        return 0.0;
    }
    samples.sort_by(|a, b| a.partial_cmp(b).unwrap());
    if samples.len() == 1 {
        return samples[0];
    }
    let k = (samples.len() - 1) as f64 * q;
    let f = k.floor() as usize;
    let c = k.ceil() as usize;
    if f == c {
        return samples[f];
    }
    samples[f] + (samples[c] - samples[f]) * (k - f as f64)
}

#[derive(Clone)]
pub struct Stats {
    pub backend: String,
    pub scenario: String,
    pub ops: u64,
    pub errors: u64,
    pub reads: u64,
    pub hits: u64,
    pub total_s: f64,
    pub mean_ms: f64,
    pub p50_ms: f64,
    pub p95_ms: f64,
    pub p99_ms: f64,
}

#[allow(clippy::too_many_arguments)]
impl Stats {
    pub fn new(
        backend: &str,
        scenario: &str,
        ops: u64,
        errors: u64,
        reads: u64,
        hits: u64,
        total_s: f64,
        mut latencies_ms: Vec<f64>,
    ) -> Self {
        let mean = if latencies_ms.is_empty() {
            0.0
        } else {
            latencies_ms.iter().sum::<f64>() / latencies_ms.len() as f64
        };
        Self {
            backend: backend.to_string(),
            scenario: scenario.to_string(),
            ops,
            errors,
            reads,
            hits,
            total_s,
            mean_ms: mean,
            p50_ms: percentile(&mut latencies_ms, 0.50),
            p95_ms: percentile(&mut latencies_ms, 0.95),
            p99_ms: percentile(&mut latencies_ms, 0.99),
        }
    }

    pub fn ops_per_sec(&self) -> f64 {
        if self.total_s > 0.0 {
            self.ops as f64 / self.total_s
        } else {
            0.0
        }
    }

    pub fn hit_ratio(&self) -> f64 {
        if self.reads > 0 {
            self.hits as f64 / self.reads as f64
        } else {
            0.0
        }
    }

    pub fn to_json(&self) -> String {
        format!(
            r#"{{"backend":"{}","scenario":"{}","ops":{},"errors":{},"reads":{},"hit_ratio":{:.3},"total_s":{:.3},"ops_per_sec":{:.1},"mean_ms":{:.4},"p50_ms":{:.4},"p95_ms":{:.4},"p99_ms":{:.4}}}"#,
            self.backend,
            self.scenario,
            self.ops,
            self.errors,
            self.reads,
            self.hit_ratio(),
            self.total_s,
            self.ops_per_sec(),
            self.mean_ms,
            self.p50_ms,
            self.p95_ms,
            self.p99_ms
        )
    }
}
